main.cpp

```diff
@@ -31,62 +32,50 @@ struct T
 
 struct ComparingMachine                              //4
 {
-    T* compare(T* a, T* b) //5
+    const T* compare(const T& a, const T& b) //5
```

Comment:
Good.  however, you changed the return type.
No one said to make the function pararmeters const...
- [ ] revise.